### PR TITLE
iproute2: update to 5.11.0 and fix unneeded dependencies

### PR DESCRIPTION
**Pull request by guidosarducci**
~~**NOTE**: _This is in anticipation of an imminent `iproute2 5.11` release, which adds key features like `libbpf` support, while this commit also adds some improvements and fixes an old dependency problem. When 5.11 drops, this PR can be quickly updated, re-tested and reviewed/merged. If there's a desire, this early version could also be used as is and updated later. Ideally, these changes should target both master and the kernel-5.4-based 21.02 release._~~

The latest iproute2 version brings various improvements and fixes:
https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/log/?qt=range&q=v5.10.0..v5.11.0

In particular, ip and tc now use libbpf as the standard way to load BPF
programs, rather than the old, limited custom loader. This allows more
consistent and featureful BPF program handling e.g. support for global
initialized variables.

Also fix a longstanding problem with iproute2 builds where unneeded DSO
dependencies are added to most utilities, bloating their installation
footprint. From research and testing, explicitly using a "--as-needed"
linker flag avoids the issue. Update accordingly and drop extra package
dependencies from Makefile.

Additional build and packaging updates include:

  - install missing development header to iproute2/bpf_elf.h
  - propagate OpenWrt verbose flag during build
  - update and refresh patches

Compile and run tested: QEMU/malta-mips32be on kernels 5.4 & 5.10.

All iproute2 packages were built and installed to the test image. Some
regression testing using ip-full and tc was successfully performed to
exercise several kmods, tc modules, and simple BPF programs.

Pinging the maintainer and others regularly committing to the package: @RussellSenior , @hauke  and @dedeckeh .

Thanks for your feedback.

### CODE REVIEW

File: package/network/utils/iproute2/Makefile [hunk 1]
```diff
@@ -8,12 +8,16 @@
...
-PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
-PKG_SOURCE_URL:=@KERNEL/linux/utils/net/iproute2
-PKG_HASH:=a54a34ae309c0406b2d1fb3a46158613ffb83d33fefd5d4a27f0010237ac53e9
+PKG_SOURCE_VERSION:=5a37254b71249bfb73d44d6278d767a6b127a2f9
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://git.kernel.org/pub/scm/network/iproute2/iproute2.git
+PKG_MIRROR_HASH:=
+#PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+#PKG_SOURCE_URL:=@KERNEL/linux/utils/net/iproute2
+#PKG_HASH:=a54a34ae309c0406b2d1fb3a46158613ffb83d33fefd5d4a27f0010237ac53e9
```
Comment (adschm) [inline package/network/utils/iproute2/Makefile:20]: Drop the comment?

Comment (guidosarducci) [reply package/network/utils/iproute2/Makefile:20]: Sure, I intend to, but left this is as placeholder/reminder about the final version (release or pre-release) being undetermined, much like the empty `PKG_MIRROR_HASH:=`. For me, the PR has big improvements, but I have no insight into when/if a 5.11 iproute2 release will drop early enough before the 21.02 release to get this in. That's one aspect of feedback I'm looking for from the maintainer/past committers.

What's your opinion? Aim to merge this pre-release and update later? Or wait and see (but for how long)? Would it be better to mark this "draft" while we decide?

**Comment by RussellSenior** [comment]
It would be interesting to see how the sizes change. Do you happen to have that info handy? I'll give it a build/run-test tomorrow.

**Comment by guidosarducci** [comment]
@RussellSenior That was a concern of mine too, but the package sizes don't change much at all. The main difference now is with package dependencies and their installation footprint. Previously `bridge`, `devlink`, `genl`, `nstat`, `rdma`, `ss` would have had unneeded dependencies on `libelf` and `libbpf`, but that's fixed now.

Only `tc` and `ip-full` now depend on libbpf (and libelf), and for reference those lib sizes are:
```
-rw-r--r-- 1 kodidev kodidev 91582 Feb 21 22:39 libbpf0_5.10.10-1_mips_24kc.ipk
-rw-r--r-- 1 kodidev kodidev 43204 Feb 21 22:39 libelf1_0.180-1_mips_24kc.ipk
```

Since few things use `libelf` and `libbpf` these days, I looked into statically linking those with `tc` and `ip-full`, but total sizes were ~30KB larger than just dynamic linking (`tc` + `ip-full` + `libbpf` +`libelf`), even with all the right build options: `-ffunction-sections -fdata-sections -flto -Wl,--gc-sections`. So in the end I decided the current setup was best.

Some numbers for reference:

**1. Build:  iproute2 5.10.0, libelf enabled (dynamic linked to all)**
```
-rw-r--r-- 1 120591 Feb 19 23:31 ip-tiny_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 170192 Feb 19 23:32 ip-full_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 150427 Feb 19 23:33 tc_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 22821 Feb 19 23:33 genl_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 23909 Feb 19 23:33 ip-bridge_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 37514 Feb 19 23:33 ss_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 7576 Feb 19 23:33 nstat_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 28700 Feb 19 23:33 devlink_5.10.0-1_mips_24kc.ipk
-rw-r--r-- 1 18178 Feb 19 23:33 rdma_5.10.0-1_mips_24kc.ipk
```
**2. Build: iproute2 5.11rc1, libelf & libbpf enabled (dynamic linked only with ip-full & tc) [CURRENT SETUP]**
```
-rw-r--r-- 1 121427 Feb 21 02:32 ip-tiny_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 168076 Feb 21 02:33 ip-full_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 148343 Feb 21 02:34 tc_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 23574 Feb 21 02:34 genl_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 24358 Feb 21 02:34 ip-bridge_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 37533 Feb 21 02:34 ss_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 7539 Feb 21 02:34 nstat_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 29253 Feb 21 02:34 devlink_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 18144 Feb 21 02:34 rdma_5.11rc1-1_mips_24kc.ipk
```
**3. Build: iproute2 5.11rc1, libelf & libbpf enabled (static linked only with ip-full & tc)**
```
-rw-r--r-- 1 121413 Feb 21 04:52 ip-tiny_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 252670 Feb 21 04:53 ip-full_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 231817 Feb 21 04:54 tc_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 23570 Feb 21 04:54 genl_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 24351 Feb 21 04:54 ip-bridge_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 37519 Feb 21 04:54 ss_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 7532 Feb 21 04:54 nstat_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 29249 Feb 21 04:54 devlink_5.11rc1-1_mips_24kc.ipk
-rw-r--r-- 1 18142 Feb 21 04:54 rdma_5.11rc1-1_mips_24kc.ipk
```

File: package/network/utils/iproute2/Makefile (hunk 1 above)
Comment (adschm) [reply package/network/utils/iproute2/Makefile:20]: Okay, I only had a quick look and didn't really catch that in the first place.
Indeed, marking as draft will be an indicator for something that is not to be merged yet (although it won't have any practical effect for the core repo where stuff is picked from PRs).
I will put the "blocked" tag, as that's how I understand the current status then?

Comment (guidosarducci) [reply package/network/utils/iproute2/Makefile:20]: That should be fine as we only want to prevent a "drive-by" merge before hearing from maintainers. The current situation is a bit strange: usually the iproute2 release trails the kernel release by only 3-4 days.

**Comment by guidosarducci** [comment]
Rebased and retested successfully after updating to the just-released iproute2 v5.11.0, so this should now be ready for review/merge.

@adschm  Could you drop the "blocked" tag and then we'll look for further comments from @RussellSenior and others. Thanks!

**Comment by RussellSenior** [comment]
I built and ran on ath79 without problems (with the caveat that I've never used genl, nstat, ss or devlink). they at least run the -V version option and return sensible results without segfaulting.

File: package/network/utils/iproute2/Makefile [hunk 2]
```diff
@@ -122,26 +125,29 @@ define Build/Configure
 endef
 
 TARGET_CFLAGS += -ffunction-sections -fdata-sections -flto
-TARGET_LDFLAGS += -Wl,--gc-sections
+TARGET_LDFLAGS += -Wl,--gc-sections -Wl,--as-needed
```
Comment (dengqf6) [inline package/network/utils/iproute2/Makefile:128]: This can be merged into the previous `-Wl` argument, i.e. `-Wl,--gc-sections,--as-needed`

**Comment by guidosarducci** [comment]
> I built and ran on ath79 without problems (with the caveat that I've never used genl, nstat, ss or devlink). they at least run the -V version option and return sensible results without segfaulting.

@RussellSenior I'm in the same boat, doing similar simple tests myself, and never "using them in anger". Having said that, at least `nstat` and `ss` have relevant default output that shows they're working. Besides that, I've also seen the same outputs from previous iproute2 versions, so no indication of anything wrong (unless they were previously wrong!).

File: package/network/utils/iproute2/Makefile (hunk 2 above)
Comment (guidosarducci) [reply package/network/utils/iproute2/Makefile:128]: Good of you to notice; however "can" and "should" are really different things. I've done that deliberately in the interest of maintainability and avoiding confusion, rather than conflate a rare linker workaround with the multi-line size optimization (`-ffunction-sections -fdata-sections -flto` / `-Wl,--gc-sections`) often used in package Makefiles.

I don't think saving 4 bytes outweighs the above.

**Comment by guidosarducci** [comment]
@adschm I talked to @RussellSenior, the package maintainer, via IRC earlier. He was OK with the changes, but mentioned he had no ability to review/approve changes. What's the best path forward in these situations?

I'm in the same position myself with another package I maintain, `bpftools`.

**Comment by adschm** [comment]
> @adschm I talked to @RussellSenior, the package maintainer, via IRC earlier. He was OK with the changes, but mentioned he had no ability to review/approve changes. What's the best path forward in these situations?
> 
> I'm in the same position myself with another package I maintain, `bpftools`.

Well, Acked-by/Reviewed-by are always helpful.

Apart from that, as you can probably deduce from my comments, packages are not my main area of activity. I won't merge this in any case.

Typical way forward would be to check git history on who has been active on this package/area. Then ping the person(s) here on GitHub ... But packages in core repo are nothing that is really fast to change in general.

**Comment by guidosarducci** [comment]
> Well, Acked-by/Reviewed-by are always helpful.

Thanks, I'll reconfirm and update the commit

> Typical way forward would be to check git history on who has been active on this package/area. Then ping the person(s) here on GitHub ... But packages in core repo are nothing that is really fast to change in general.

Indeed, I've regularly done just that, including with this PR. Unfortunately to limited effect .

**Comment by RussellSenior** [comment]
Acked-by: Russell Senior <russell@personaltelco.net>

File: package/network/utils/iproute2/Makefile [hunk 3]
```diff
@@ -122,26 +125,29 @@ define Build/Configure
...
+	$(if $(findstring c,$(OPENWRT_VERBOSE)),V=1,V='')
 
 define Build/Compile
 	+$(MAKE_VARS) $(MAKE) $(PKG_JOBS) -C $(PKG_BUILD_DIR) $(MAKE_FLAGS)
 endef
 
 define Build/InstallDev
-	$(INSTALL_DIR) $(1)/usr/include
+	$(INSTALL_DIR) $(1)/usr/include/iproute2
+	$(CP) $(PKG_BUILD_DIR)/include/bpf_elf.h $(1)/usr/include/iproute2
```
Comment (hauke) [inline package/network/utils/iproute2/Makefile:150]: Why did you change the files which are copied over?

Comment (guidosarducci) [reply package/network/utils/iproute2/Makefile:150]: It's a missing development header (bpf_elf.h), previously overlooked, that helps with BPF map definitions typically used with tc. It's mentioned in the commit message too (though not as detailed) and the existing headers are still copied. What was the concern exactly?

Comment (guidosarducci) [reply package/network/utils/iproute2/Makefile:150]: To avoid any doubt, it bears mentioning that the previous install directories and header files do not change; they are still installed to the same locations with:
```
$(CP) $(PKG_BUILD_DIR)/include/{libgenl,libnetlink}.h $(1)/usr/include/
```

The change to `$(INSTALL_DIR) $(1)/usr/include/iproute2` only creates an additional subdirectory for the `bpf_elf.h` header, supporting the `tc` BPF loader. This header install location is standard with upstream `iproute2` and also used in some other distro packages e.g. Ubuntu, Alpine, Arch. There's also some [related Debian discussion](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=904122#35). 

Hope that clarifies things.

Comment (hauke) [reply package/network/utils/iproute2/Makefile:150]: thanks for the clarification.

**Comment by hauke** [comment]
Thank you for the pull request, it was added to master.

**Comment by guidosarducci** [comment]
@hauke Thank you for reviewing and merging. Based on concerns about the growing size of the `tc` package, I've also implemented and tested a `tc-tiny` variant of `iproute2` in https://github.com/openwrt/openwrt/pull/3961.
